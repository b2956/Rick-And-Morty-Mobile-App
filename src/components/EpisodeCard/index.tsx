import React from 'react';
import styled from 'styled-components/native';

export interface IEpisodeProps {
    id: string
    name: string
    air_date: string
    episode: string
    characters: ICharacter[]
}

interface ICharacter {
    __typename: string,
    name: string
}

const Wrapper = styled.View`
    width: 95%;
    height: 200px;
    background-color: #606268;
    margin-bottom: 10px;
    align-self: center;
    border-radius: 10px;
    flex-direction: column;
    overflow: hidden;
    padding-left: 15px;
    padding-right: 15px;
    justify-content: center;
`;

const Title = styled.Text`
    font-size: 20px;
    color: #fff;
    font-weight: 700;
    margin-bottom: 10px;
`;

const EpisodeInfoWrapper = styled.View`
    width: 100%;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
`;

const EpisodeInfoBlock = styled.View`
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
`;

const GreyText = styled.Text`
    color: rgb(158, 158, 158);
    font-size: 13px;
    font-weight: 400;
`;

const WhiteText = styled.Text`
    color: #fff;
    margin-bottom: 10px;
    font-size: 13px;
`;

const EpisodeCard = ({id, name, air_date, episode, characters}: IEpisodeProps) => {
    const charactersList = characters.reduce((reducer, item, index) => {
        if(index === 0) return item.name;
        if(reducer.length >= 150 && index === characters.length - 1) return `${reducer} ...`;
        if(reducer.length >= 150) return `${reducer}`;
        return `${reducer} | ${item.name}`;
    }, '');

    return (
        <Wrapper>
            <Title>{name}</Title>
            <EpisodeInfoWrapper>
                <EpisodeInfoBlock>
                    <GreyText>Air Date:</GreyText>
                    <WhiteText>{air_date}</WhiteText>
                </EpisodeInfoBlock>
                <EpisodeInfoBlock>
                    <GreyText>Episode:</GreyText>
                    <WhiteText>{episode}</WhiteText>
                </EpisodeInfoBlock>
            </EpisodeInfoWrapper>
            <GreyText>Characters present:</GreyText>
            <WhiteText>{charactersList}</WhiteText>
        </Wrapper>
    )
};

export default EpisodeCard;