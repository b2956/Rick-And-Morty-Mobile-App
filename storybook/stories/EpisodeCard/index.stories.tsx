import React from 'react';
import {storiesOf} from '@storybook/react-native';

import EpisodeCard, {IEpisodeProps} from '../../../src/components/EpisodeCard';

const EpisodeProps: IEpisodeProps = {
    air_date: 'December 2, 2013',
    episode: 'S01E01',
    id: '1',
    name: 'Pilot',
    characters: [
        {
            __typename: 'Character',
            name: 'Rick'
        },
        {
            __typename: 'Character',
            name: 'Morty'
        },
    ]
};

storiesOf('EpisodeCard', module)
    .add('Default', () => <EpisodeCard {...EpisodeProps} />);