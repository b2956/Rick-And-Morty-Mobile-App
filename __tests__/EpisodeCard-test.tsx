import React from 'react';
import renderer from 'react-test-renderer';

import EpisodeCard, {IEpisodeProps} from '../src/components/EpisodeCard'; 

describe('Character Card', () => {
    const EpisodeProps: IEpisodeProps = {
        air_date: 'December 2, 2013',
        episode: 'S01E01',
        id: '1',
        name: 'Pilot',
        characters: [
            {
                __typename: 'Character',
                name: 'Rick'
            },
            {
                __typename: 'Character',
                name: 'Morty'
            },
        ]
    };

    it('Should render the Episode Card component correctly', () => {
        const tree = renderer.create(<EpisodeCard {...EpisodeProps} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});