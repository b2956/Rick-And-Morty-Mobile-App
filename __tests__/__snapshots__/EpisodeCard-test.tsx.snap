// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Card Should render the Episode Card component correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "center",
        "backgroundColor": "#606268",
        "borderRadius": 10,
        "flexDirection": "column",
        "height": 200,
        "justifyContent": "center",
        "marginBottom": 10,
        "overflow": "hidden",
        "paddingLeft": 15,
        "paddingRight": 15,
        "width": "95%",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 20,
          "fontWeight": "700",
          "marginBottom": 10,
        },
      ]
    }
  >
    Pilot
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "column",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(158, 158, 158)",
              "fontSize": 13,
              "fontWeight": "400",
            },
          ]
        }
      >
        Air Date:
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 13,
              "marginBottom": 10,
            },
          ]
        }
      >
        December 2, 2013
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "column",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(158, 158, 158)",
              "fontSize": 13,
              "fontWeight": "400",
            },
          ]
        }
      >
        Episode:
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 13,
              "marginBottom": 10,
            },
          ]
        }
      >
        S01E01
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "rgb(158, 158, 158)",
          "fontSize": 13,
          "fontWeight": "400",
        },
      ]
    }
  >
    Characters present:
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 13,
          "marginBottom": 10,
        },
      ]
    }
  >
    Rick | Morty
  </Text>
</View>
`;
